### PR TITLE
[Scheduler] Support per task/actor PlacementGroupSchedulingStrategy

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
This PR adds per task/actor scheduling strategy and currently the only strategy are PlacementGroupSchedulingStrategy and DefaultSchedulingStrategy.

Going forward, people should use `scheduling_strategy=PlacementGroupSchedulingStrategy` to define placement group for actor/task. The old way will be deprecated.

## Related issue number
#18524

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jjyao** [comment]
No need to review now. I just publish the PR to generate a wheel for testing.

File: python/ray/worker.py
Comment (jjyao) [inline python/ray/worker.py:2193]: @ericl @rkooo567 Should we call it scheduling_strategy to be more coherent with placement group strategy?

Comment (rkooo567) [reply python/ray/worker.py:2193]: Yeah +1. (Also using the same term for values)

**Comment by jjyao** [comment]
Move round robin fix to a separate PR since this PR is getting too large.

**Comment by ericl** [comment]
Ready for review?

**Comment by jjyao** [comment]
@ericl You can do the high level review now: api, protobuf changes. The things left are just polish: tests, comments.

File: cpp/src/ray/runtime/task/native_task_submitter.cc
Comment (ericl) [inline cpp/src/ray/runtime/task/native_task_submitter.cc:66]: Could we split out the placement group strategy refactor in a separate PR this can be based on?

File: python/ray/actor.py
Comment (ericl) [inline python/ray/actor.py:488]: Update pydoc for scheduling_strategy, and comment that placement_group* are deprecated?

Comment (ericl) [reply python/ray/actor.py:488]: We should also make a (maybe another separate) PR refactoring internal code to not use placement_group, and updating the docs.

**Comment by rkooo567** [comment]
Maybe consider splitting up the PR? I think we can separate out placement group strategy code.

**Comment by jjyao** [comment]
Separated them out. This PR only contains PlacementGroupSchedulingStrategy.

**Comment by rkooo567** [comment]
I feel like deprecating the API right away is too drastic change? Since it will break all placement group use cases. 

Why don't we print the warning first, and deprecate it when we release 2.0? (or maybe in a couple versions later, like 1.12?)

**Comment by jjyao** [comment]
@ericl @rkooo567 It's ready for more reviews.

File: python/ray/_raylet.pyx
Comment (rkooo567) [inline python/ray/_raylet.pyx:1453]: Can you add an error message here?

File: python/ray/actor.py
Comment (rkooo567) [inline python/ray/actor.py:303]: Consider adding the type annotation?

File: python/ray/actor.py
Comment (rkooo567) [inline python/ray/actor.py:289]: This means we allow to create
```
@ray.remote(scheduling_strategy="blah blah")
def f():
    
```

right?

File: python/ray/tests/test_scheduling_strategy.py
Comment (rkooo567) [inline python/ray/tests/test_scheduling_strategy.py:11]: Why don't we just add all strategy tests to `test_scheduling` instead of making a new test suite?

File: python/ray/_raylet.pyx
Comment (rkooo567) [inline python/ray/_raylet.pyx:1436]: Is `clear_scheduling_strategy ` a protobuf method, or setting the nullptr for the `c_scheduling_strategy `?

I think it is more like at the end of this method, is `c_scheduling_strategy ` guaranteed to be not null? (can you add assert?)

**Review by rkooo567 — COMMENTED** [review]
Generally looks good to me. For docstring, although they are internal facing, can we add type annotation and better docstring? (I think we should iteratively improve this part in this way)

File: python/ray/_raylet.pyx
Comment (jjyao) [reply python/ray/_raylet.pyx:1436]: yes, it's a protobuf method: https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#oneof

c_scheduling_strategy can be "null" (ONEOF_NOT_SET) indicating we want to use the default scheduling strategy.

File: src/ray/protobuf/common.proto
Comment (ericl) [inline src/ray/protobuf/common.proto:233]: Prefer not to renumber?

File: python/ray/actor.py
Comment (rkooo567) [reply python/ray/actor.py:292]: Hmm can we only disallow this to the placement group? I think it makes no sense to set this strategy for the placement group.

Comment (jjyao) [reply python/ray/actor.py:292]: @ericl, Sang's suggesting that PlacementGroupSchedulingStrategy (not other SchedulingStrategy) can only be set through `Options` not decorator.

So @ray.remote(scheduling_strategy=PlacementGroupSchedulingStrategy(...)) is not allowed.

WDYT?

File: python/ray/remote_function.py
Comment (jjyao) [reply python/ray/remote_function.py:319]: Yes, since we can only have one scheduling_strategy, once you set it to "SPREAD", there will be no placement group.

**Comment by ericl** [comment]
We need to be able to set it in the decorator, especially to disable
placement groups (set to None).

**Comment by rkooo567** [comment]
@ericl 

> We need to be able to set it in the decorator, especially to disable
placement groups (set to None).

Is it for the capturing child tasks? Why not just doing this in options?

I think it is generally not a good practice to set placement group to tasks/actors decorator because those methods are not usable as soon as placement group is removed. Currently, if pg is removed, and if you schedule tasks, it hangs (it will be fixed once infeasible task exception is supported, but it is still weird you have tasks/actors that always fail).

If setting None is necessary for enforcing "never be used by placement group policy", why don't we only allow to set None for that purpose and raise an exception otherwise?

**Comment by jjyao** [comment]
@ericl @rkooo567 It's ready for more reviews. Also we still have the open discussion about whether we should allow setting PlacementGroupSchedulingStrategy in the decorator.

File: python/ray/actor.py
Comment (ericl) [reply python/ray/actor.py:293]: I don't see a strong reason to disallow it. You might create a placement group in a function and define nested remote functions closing over the placement group.

It doesn't cause any issues not already present in placement groups, and would be a special case to disallow, so I vote to allow it (default decision for all args).

File: python/ray/util/scheduling_strategies.py
Comment (ericl) [inline python/ray/util/scheduling_strategies.py:46]: Comment that this disables any placement group capture.

File: python/ray/util/scheduling_strategies.py
Comment (ericl) [inline python/ray/util/scheduling_strategies.py:25]: Is there a reason to prefer JSON instead of just pickling this class for Ray client? It seems to add quite a bit of cruft.

File: python/ray/worker.py
Comment (ericl) [inline python/ray/worker.py:2218]: I don't think we should recommend two ways to specify DEFAULT, how about just the string?

**Comment by ericl** [comment]
Overall LGTM, but since this is a large PR it would be great to get a second reviewer (cc @rkooo567 ).

File: python/ray/actor.py
Comment (rkooo567) [reply python/ray/actor.py:293]: Got it. It sounds reasonable. I think in this case, I'd like to at least make sure the documentation states its behavior clearly. 

@jjyao can you update the documentation to make it clear this won't work if placement group is removed?

File: python/ray/_raylet.pyx
Comment (rkooo567) [inline python/ray/_raylet.pyx:1463]: Minor, but we can do `RAY_CHECK(c_scheduling_strategy != nullptr)`?

File: python/ray/util/scheduling_strategies.py
Comment (rkooo567) [inline python/ray/util/scheduling_strategies.py:45]: Commenting what's the default behavior?

**Review by rkooo567 — APPROVED** [review]
LGTM after comments are addressed

File: python/ray/actor.py
Comment (jjyao) [reply python/ray/actor.py:293]: I feel a better place to state the behavior is PG doc not the code comment here.

File: python/ray/_raylet.pyx
Comment (jjyao) [reply python/ray/_raylet.pyx:1463]: Added the check to core_worker.cc so it benefits all languages.

**Review by ericl — APPROVED** [review]

**Pull request merged**
